default/python/AI/ShipDesignAI.py

```diff
@@ -2108,3 +2120,39 @@ def _can_build(design, empire_id, pid):
     :return: bool
     """
     return design.productionLocationForEmpire(empire_id, pid)
+
+
+def recursive_dict_diff(dict_new, dict_old, dict_diff, diff_level_threshold=0):
+    """Find the entries in dict_new that are not present in dict_old and store them in dict_diff.
+
+    Example usage:
+    dict_a = {1:2, 2: {2: 3, 3: 4}}
+    dict_b = {2: {2: 3, 3: 3}}
+    diff = {}
+    recursive_dict_diff(dict_a, dict_b, diff)
+    --> diff = {1:2, 2:{3:4}}
+
+    :param dict_new:
+    :type dict_new: dict
+    :param dict_old:
+    :type dict_old: dict
+    :param dict_diff: Difference between dict_old and dict_new, modified and filled within this function
```

Comment:
For me recursive call with additional accumulator looks more complex then simple recursion.